packages/dds/tree/src/changeset/visit.ts

```diff
@@ -0,0 +1,209 @@
+ * we want to ensure that visitors visit changes in an order that guarantees all changes are describable in terms
+ * of some position in the tree. This means that we need to detach content bottom-up and attach content top-down.
+ * Note that while the attach positions are expressed top-down, there is still a bottom-up spirit to building trees
+ * that are being inserted.
+ *
+ * The second challenge, is that of the inability of the visitor to move-in content that has yet to be moved-out.
+ * This leads to a two-pass algorithm, but there are two degrees for freedom to consider:
+ * 1) Whether inserts should be performed in the first pass whenever possible (some are not: inserts below a move-ins
+ *   for which we have not yet seen the matching move-out).
+ * Pros: The path above the insertion point is walked once instead of twice
+ * Cons: The paths within the inserted content risk being walked twice instead of once (once for building the content,
+ * once for traversing the tree to reach move-in marks in the second phase).
+ *
+ * 2) Whether move-ins for which we have the move-out content should be performed in the first pass.
+ * Pros: The path above the move-in point is walked once instead of twice
+ * Cons: We now have to record which of the move-ins we did not perform in the first pass. We could build a trie of
+ * those to reduce the amount of sifting we have to do on the second pass.
+ *
+ * The presence of a move table, which lists the src and dst paths for each move, could be leveraged to make some of
+ * these option more efficient:
+ * - If inserts are allowed in the first pass and move-ins are not allowed in the first pass, then the move table
+ *   describes exactly which parts of the delta need applying in the second pass.
+ * - If inserts and move-ins are allowed in the first pass then having a boolean flag for each entry in the move table
+ *   that describes whether the move has been attached, or having a set for that describes which entries remain, would
+ *   describe which parts of the delta  need applying in the second pass.
+ *
+ * Current implementation:
+ * - Performs inserts in the first pass
+ * - Does not perform move-ins in the first pass
+ * - Skips the second pass if no move-outs were encountered in the first pass
+ * - Does not leverage the move table
+ *
+ * Future work:
+ * - Take a path to the subtree of interest (or some other description of the regions of interest)
+ * - Leverage move table when it gets added to Delta
+*/
+
+/**
+ * Crawls the given `delta`, calling `visitor`'s callback for each change encountered.
+ * Each successive call to the visitor callbacks assumes that the change described by earlier calls have been applied
+ * to the document tree. For example, for a change that deletes the first and third node of a field, the visitor calls
+ * will pass indices 0 and 1 respectively.
+ * @param delta - The delta to be crawled.
+ * @param visitor - The object to notify of the changes encountered.
```

Comment:
We should also be clear about what is preserved from the original changes in Deltas.

This might not be the place to document it (maybe is a property of who produced the delta), but it does interact with whats covered here:

Is the order changes are visited allowed to be meaningful? (Ex: is it ok for deltas to forget the order of the changes they are made of as long as you can produce indexes which are correct when they are visited).

Do we guarantee all the changes made from the edits are visited, or can some be omitted due to squashing (ex: an insert that ended up getting delegated could be skipped)?

Is coalescing of adjacent edits allowed? (Ex: if we had range edits here, could two inserts be merged?)?

If someone does a "rotate" using a move (moves one end of the trait to the other end), do we guarantee that we emit that move, or are we allowed to have the delta do the other move that leads to the same outcome (move the other half of the trait instead)?

We have had users of experimental/tree try and recover some of these kinds of editing details from its deltas and have issues around these: it would be good to set expectations in this area (even if we just do so by documenting that these things are an implementation detail of the creator of the delta instead of calling out them explicitly here)

Reply:
I agree it would be good to have these things documented somewhere. I think the converter function on ChangeFamilly would be a good place. I'll make sure we do that.

I did add something in delta.ts about the "an insert that ended up getting deleted" example so it's clear that Delta instances don't represent such things.